### PR TITLE
fenics-basix: update for v0.7

### PR DESCRIPTION
**Pull request by garth-wells**
Updates basix and py-basix for the v0.7.0 release.

### CODE REVIEW

**Comment by jhale** [comment]
@spackbot-app I approve the PR.

**Review by chrisrichardson — APPROVED** [review]

**Review by tldahlgren — DISMISSED** [review]
Confirmed the two version sha256, rest LGTM

File: var/spack/repos/builtin/packages/py-fenics-basix/package.py [hunk 1]
```diff
@@ -15,22 +15,25 @@ class PyFenicsBasix(PythonPackage):
...
+    depends_on("python@3.8:", when="@0.7.0:", type=("build", "run"))
     depends_on("py-setuptools@42:", type="build")
-    depends_on("py-setuptools@40:", type="build")
     depends_on("py-numpy@1.21:", type=("build", "run"))
-    depends_on("cmake@3.19:", type="build")
-    depends_on("python@3.7:", type=("build", "run"))
-    depends_on("py-pybind11@2.9.1:", type="build")
-    depends_on("py-pybind11@2.2.4:", type="build")
+    depends_on("py-pybind11@2.9.1:", when="@:0.6.0", type="build")
+    depends_on("py-nanobind@1.5.1:", when="@0.7.0:", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fenics-basix/package.py]: I don't see any mention of nanobind in 0.7, but I do see pybind11. Maybe this is only for 0.8+?

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-fenics-basix/package.py [hunk 2]
```diff
@@ -15,22 +15,25 @@ class PyFenicsBasix(PythonPackage):
...
 
     depends_on("fenics-basix@main", type=("build", "run"), when="@main")
+    depends_on("fenics-basix@0.7.0", type=("build", "run"), when="@0.7.0")
     depends_on("fenics-basix@0.6.0", type=("build", "run"), when="@0.6.0")
     depends_on("fenics-basix@0.5.1", type=("build", "run"), when="@0.5.1")
     depends_on("fenics-basix@0.4.2", type=("build", "run"), when="@0.4.2")
 
+    depends_on("cmake@3.19:", type="build")
+
+    depends_on("python@3.7:", when="@:0.6.0", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fenics-basix/package.py]: Could remove this line, Spack only supports Python 3.7+ anyway

File: var/spack/repos/builtin/packages/py-fenics-basix/package.py [hunk 3]
```diff
@@ -15,22 +15,25 @@ class PyFenicsBasix(PythonPackage):
...
+    depends_on("python@3.8:", when="@0.7.0:", type=("build", "run"))
     depends_on("py-setuptools@42:", type="build")
-    depends_on("py-setuptools@40:", type="build")
     depends_on("py-numpy@1.21:", type=("build", "run"))
-    depends_on("cmake@3.19:", type="build")
-    depends_on("python@3.7:", type=("build", "run"))
-    depends_on("py-pybind11@2.9.1:", type="build")
-    depends_on("py-pybind11@2.2.4:", type="build")
+    depends_on("py-pybind11@2.9.1:", when="@:0.7", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fenics-basix/package.py:36]: Missing dep on py-scikit-build-core

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-fenics-basix/package.py (hunk 3 above)
Comment (garth-wells) [reply var/spack/repos/builtin/packages/py-fenics-basix/package.py:36]: `py-scikit-build-core` is only used for 'convenience' pip install that builds the C++ library and the python interface in one step. It isn't used by Spack.

The Spack packages separate the C++ library and the and the Python wrappers, and don't require `py-scikit-build-core`. The Python requirements are listed in `python/` directory. Python requirements in the top level directory aren't relevant for the Spack packages.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-fenics-basix/package.py:36]: Ah, can you add a comment like:
```python
# python/pyproject.toml
```
just so I remember where to look for deps when I review next time?

**Comment by garth-wells** [comment]
Closing as py-cmake, which py-nanobind depends on, fails to build. Will re-open shortly.

**Comment by adamjstewart** [comment]
I've always been confused by py-cmake. It's supposed to just be a wrapper around cmake, but it never works for me in Spack. Wondering if we should just get rid of py-cmake and py-ninja entirely, or if it's possible to fix them somehow.

**Comment by garth-wells** [comment]
> I've always been confused by py-cmake. It's supposed to just be a wrapper around cmake, but it never works for me in Spack. Wondering if we should just get rid of py-cmake and py-ninja entirely, or if it's possible to fix them somehow.

I think removal is preferable - otherwise all the intricacies in the cmake package need to be duplicated in the py-cmake package file.

**Comment by garth-wells** [comment]
@adamjstewart I think this PR is ready now.

File: var/spack/repos/builtin/packages/py-fenics-basix/package.py (hunk 3 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-fenics-basix/package.py:36]: Just waiting on this comment

Comment (garth-wells) [reply var/spack/repos/builtin/packages/py-fenics-basix/package.py:36]: Done.

File: var/spack/repos/builtin/packages/py-fenics-basix/package.py [hunk 4]
```diff
@@ -15,22 +15,25 @@ class PyFenicsBasix(PythonPackage):
...
     version("0.5.1", sha256="69133476ac35f0bd0deccb480676030378c341d7dfb2adaca22cd16b7e1dc1cb")
     version("0.4.2", sha256="a54f5e442b7cbf3dbb6319c682f9161272557bd7f42e2b8b8ccef88bc1b7a22f")
 
     depends_on("fenics-basix@main", type=("build", "run"), when="@main")
+    depends_on("fenics-basix@0.7.0", type=("build", "run"), when="@0.7.0")
     depends_on("fenics-basix@0.6.0", type=("build", "run"), when="@0.6.0")
     depends_on("fenics-basix@0.5.1", type=("build", "run"), when="@0.5.1")
     depends_on("fenics-basix@0.4.2", type=("build", "run"), when="@0.4.2")
 
+    depends_on("cmake@3.19:", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fenics-basix/package.py]:
```suggestion
    # See python/CMakeLists.txt
    depends_on("cmake@3.16:", type="build")
```

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**